### PR TITLE
Helical filament

### PR DESCRIPTION
**Pull request by brisvag**
Simple subclass of `Spline3D` which adds a new method for sampling a helical filament. Not sure if I should keep this more general for this package, but for now this is almost identical to how it was in `tomogeom`.

### CODE REVIEW

File: src/morphosamplers/helical_filament.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from scipy.spatial.transform import Rotation
+import numpy as np
+
+from .spline import Spline3D
+
+
+class HelicalFilament(Spline3D):
+
+    def sample_helical(self, rise, twist=0, radius=0, cyclic_symmetry_order=1, twist_offset=0, degrees=True):
```
Comment (alisterburt) [inline src/morphosamplers/helical_filament.py:9]: would be good to have some definitions of the meaning of the parameters here, something like

- rise is the spacing between samples along the helical axis (the spline)
- twist is the rotational offset around the helical axis between samples in degrees
- radius is the displacement of samples from the helical axis
- cyclic symmetry order allows for cyclic symmetrisation of each sample around the helical axis
- twist offset is an offset of the 'twist' parameter applied to all samples

I would probably remove the option to change between degrees/radians given that you don't pass it through (see below) and it's yet another kwarg here

File: src/morphosamplers/helical_filament.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        positions = []
+        orientations = []
+        base_positions = self.sample(separation=rise)
+        base_orientations = self.sample_orientations(separation=rise)
+
+        # twist around the z acis by the given angle
+        twist_angles = twist_offset + (np.arange(len(base_positions)) * twist)
+        Rz = Rotation.from_euler('z', angles=twist_angles, degrees=True)
```
Comment (alisterburt) [inline src/morphosamplers/helical_filament.py:17]: you don't pass degrees down from the kwargs here

**Review by alisterburt — COMMENTED** [review]
Looking good, thanks for the PR! Definitely want to have something like this in, a few thoughts... 

small things:
- if the object is a `HelicalFilament` maybe just `sample` is a better name for the method?
- docstring would be good, suggestions for parameter definitions below

broader questions:
The single sampling method you have here covers what seem to me like two distinct sampling 'cases' which might be easier to reason about separately
- sampling positions along the helical axis, samples are oriented with Z along the helical axis 
- sampling positions offset from the helical axis, samples are oriented with Z perpendicular to the helical axis 

It's possible that my second point on the orientation of the Z axis is just a matter of expectation, whether the Z is along the helical axis (what your sampling method does) or perpendicular to the surface... the reason I have this expectation is that I've seen biological filaments where the subunits have an additional Cn symmetry perpendicular to the helical axis, sampling in the way I suggest would allow this to be leveraged more easily in refinement packages.

Concrete suggestions:
I think I would prefer to see/use something like this, with spline objects being used in the sampling methods rather than this being a subclass

```python
@evented
class HelicalFilament3D(BaseModel):
    rise: float
    twist: float
    backbone_points: np.ndarray  # (n, 3)

    def sample_helical_axis(twist_offset: float) -> np.ndarray:
        ...
    
    def sample_offset_points(radius, twist_offset, cyclic_symmetry_order):
        ...
```

Thoughts?

**Comment by brisvag** [comment]
>     * if the object is a `HelicalFilament` maybe just `sample` is a better name for the method?

I started that way, but overriding `sample` is a problem because it's also used internally by the superclass. One approach I initially thought to use was to have `HelicalFilament` be a wrapper rather than a subclass, but I didn't really like it as much. Still, it's an option.

Also, I calling it differently differentiates from the "base" method which only returns positions or only orientations. Since here they're influencing each other, it seemed like a waste to do the work twice in two separate methods.

Basically, I see this as a utility wrapper around the base implementation of `Spline3D`.

 
>     * docstring would be good, suggestions for parameter definitions below

Yup, will add as soon as we settle on how this will look like!
 
> broader questions: The single sampling method you have here covers what seem to me like two distinct sampling 'cases' which might be easier to reason about separately

Yeah this is kinda the same discussion we had in tomogeom. My view was that it's actually the same case, just that one is the special case of `radius=0`.

> the reason I have this expectation is that I've seen biological filaments where the subunits have an additional Cn symmetry perpendicular to the helical axis, sampling in the way I suggest would allow this to be leveraged more easily in refinement packages.

Isn't that also true for helices with radius=0 though?

The way I proposed to "solve" this the first time around was to have a utility function that rolls the axes of a particle (equivalent to rotating around the particle's (1, 1, 1) by 120 degrees). We could have a kwarg that allows to choose how to orient the particles.

> Concrete suggestions: I think I would prefer to see/use something like this, with spline objects being used in the sampling methods rather than this being a subclass
> 
> ```python
> @evented
> class HelicalFilament3D(BaseModel):
>     rise: float
>     twist: float
>     backbone_points: np.ndarray  # (n, 3)
> 
>     def sample_helical_axis(twist_offset: float) -> np.ndarray:
>         ...
>     
>     def sample_offset_points(radius, twist_offset, cyclic_symmetry_order):
>         ...
> ```
> 
> Thoughts?

I don't find those methods very intuitive... Also, I kinda like about the current implementation that the only static thing is the definition of the spline, and you can resample it on the fly by changing the kwargs... But yeah, I also see the benefit of a "static" definition of the helix... @kevinyamauchi, comments?

**Comment by alisterburt** [comment]
> I started that way, but overriding sample is a problem because it's also used internally by the superclass. One approach I initially thought to use was to have HelicalFilament be a wrapper rather than a subclass, but I didn't really like it as much. Still, it's an option.

Yeah, I think I prefer the wrapper for all the usual 'composition over inheritance' reasons but it can go either way

> Also, I calling it differently differentiates from the "base" method which only returns positions or only orientations. Since here they're influencing each other, it seemed like a waste to do the work twice in two separate methods.

I think this is partly why I think having a composite object is more natural - I wouldn't expect `HelicalFilament.sample()`to not be sampling the helix

> Yeah this is kinda the same discussion we had in tomogeom. My view was that it's actually the same case, just that one is the special case of radius=0.

Uh-oh, don't remember the details but sounds like we're both consistent in disagreeing about it at least 😆 
Given that a radius 0 filament can't actually exist I still feel that it's a special case because it represents something fundamentally different, points along a helical path/axis rather than subunits of a particular instance of a helix

> Isn't that also true for helices with radius=0 though?

Kinda same point as above - when 'cropping a helix with radius=0' you're not actually trying to land at a particular subunit on the helix, just ensuring that all samples have the correct relative spacing/rotational offset between them - it's why I think it's worth having two cases. I'm sure there are filaments with a C2 along z in the subunit though too

**Comment by brisvag** [comment]
Lol, I remember answering the same thing last time :P But yeah, they do exist, I'm currently working with a filament which is essentially a stack of subunits with no offset from the filament axis. The actual final subunits. There's no C2 symmetry in this case, but I can imagine that could happen.

**Comment by alisterburt** [comment]
but then it's not a 'radius' of the filament, it's some property of the helical path? an `offset`?

**Comment by kevinyamauchi** [comment]
Hey @brisvag ! I'm happy to include this. It looks like the the CI failure is real. Happy to get this in once the tests are passing.

**Comment by brisvag** [comment]
CI issue was just a dependency problem with pydantic, which I already solved in #28 by requiring `pydantic<2`!

**Pull request merged**